operators/pkg/examagent/instance.go

```diff
@@ -0,0 +1,289 @@
+	w.WriteHeader(http.StatusCreated)
+	fmt.Fprint(w, httpPageStartingUp)
+}
+
+// HandlePut handles the PUT request for a InstanceAdapter api call.
+func (ih *InstanceHandler) HandlePut(w http.ResponseWriter, r *http.Request, log logr.Logger) {
+	if err := Options.CheckAllowedIP(r.Header.Get("X-Forwarded-For")); err != nil {
+		log.Error(err, "unauthorized")
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprint(w, "Forbidden")
+		return
+	}
+
+	// get Instance from the request
+	adapter, err := InstanceAdapterFromRequest(r, log)
+	if err != nil {
+		log.Error(err, "cannot parse request")
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprint(w, "Bad request")
+		return
+	}
+
+	instance := EmptyInstanceFromRequest(r)
+	log = log.WithValues("instance", instance.Name)
+
+	op, err := ctrl.CreateOrUpdate(r.Context(), ih.Client, instance, func() error {
+		instance.Spec = InstanceSpecFromAdapter(&adapter)
+		return nil
+	})
+
+	log = log.WithValues("operation", op)
+
+	if err != nil {
+		log.Error(err, "failed performing operation")
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "Instance %s cannot be %s", instance.Name, op)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+
+	switch op {
+	case ctrlutil.OperationResultCreated:
+		w.WriteHeader(http.StatusCreated)
+	case ctrlutil.OperationResultUpdated:
+		w.WriteHeader(http.StatusOK)
+	case ctrlutil.OperationResultNone:
+		w.WriteHeader(http.StatusNotModified)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	if err := json.NewEncoder(w).Encode(AdapterFromInstance(instance)); err != nil {
+		fmt.Fprint(w, op)
+		log.Error(err, "operation complete but cannot encode instance")
+		return
+	}
+	log.Info("success")
+}
+
+// HandleGetAll handles the GET request for all the instances.
+func (ih *InstanceHandler) HandleGetAll(w http.ResponseWriter, r *http.Request, log logr.Logger) {
+	var instances clv1alpha2.InstanceList
+	// get all the instances in the namespace
+	if err := ih.Client.List(r.Context(), &instances, client.InNamespace(Options.TemplatesNS)); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Error retrieving instances")
+		log.Error(err, "error retrieving instances")
+		return
+	}
+
+	adapters := make([]InstanceAdapter, len(instances.Items))
+	for i := range instances.Items {
+		adapters[i] = *AdapterFromInstance(&instances.Items[i])
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(adapters); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Internal server error")
+		log.Error(err, "cannot encode instances")
+		return
+	}
+}
+
+// GetInstanceIDFromRequest returns the instance id from the request.
+func GetInstanceIDFromRequest(r *http.Request) string {
+	InstanceEP := path.Join(Options.BasePath, InstanceAdapterEndpoint) + "/"
+	instID := strings.Replace(r.URL.Path, InstanceEP, "", 1)
+	return instID
+}
```

Comment:
Would move this after the delete method, to segregate receiver and non-receiver methods.